stan/math/prim/fun/interp1.hpp

```diff
@@ -1,45 +1,36 @@
 #ifndef STAN_MATH_PRIM_FUN_INTERP1_HPP
 #define STAN_MATH_PRIM_FUN_INTERP1_HPP
 
-#include <stan/math/prim/meta.hpp>
+//#include <stan/math/prim/fun/sort_asc.hpp>
+
 using namespace std;
 namespace stan {
     namespace math {
-    /**
-     * Return the one-dimensional interpolation of the second argument of points corresponding to the first argument.
-     * @tparam T_true type of the true argument
-     * @tparam T_false type of the false argument
-     * @param c Boolean condition value.
-     * @param y_true Value to return if condition is true.
-     * @param y_false Value to return if condition is false.
-     */
-    inline Eigen::VectorXd
-    interp1(const Eigen::VectorXd& xData, const Eigen::VectorXd& yData, const Eigen::VectorXd& xVector)
-    {
-
-        Eigen::VectorXd yVals;
-        int size_x = xData.size();
-        int size_xData = xData.size();
+        template <typename T>
+        inline std::vector<double> interp1(const std::vector<double>& xData, const std::vector<double>& yData, const std::vector<T>& xTest)
```

Comment:
The next step here is getting gradients.

So back to the autodiff comment, but the output type scalar type should be the least-upper bound of the input scalar types (I think LUB of the scalar types is the way to talk about this, anyway, could be wrong).

Since you have a `T`, we have to think about this.

In Stan the `T` might be instantiated to `double`, `var`, `fvar<double>`, or some other higher order autodiff type.

If the input is an autodiff type, then the output will be an autodiff type too. So the output type in this case should be `std::vector<stan::math::return_type_t<double, double, T>>` which just simplifies to `std::vector<T>` if `T` is any of the types we might expect.

This also means the temporaries in this calculation should be `T`.

Reply:
Add this code as the file `test/unit/math/rev/fun/interp1_test.cpp`:

```
#include <stan/math/rev.hpp>
#include <gtest/gtest.h>
#include <limits>

TEST(AgradRev, interp1_gradient) {
  using stan::math::var;
  using stan::math::interp1;

  std::vector<double> x = { 0.1, 1.2 };
  std::vector<double> y = { 1.0, -2.0 };

  var x_int = 0.7;
  std::vector<var> x_int_vec = { x_int };

  std::vector<var> y_int_vec = interp1(x, y, x_int_vec);
  var y_int = y_int_vec[0];

  y_int.grad();

  // x_int.adj() should contain dy_int/dx_int                                                                                                                                
  EXPECT_FLOAT_EQ(-3.0 / 1.1, x_int.adj());
}
```

I tried to push to the repo but it said I didn't have write access. Can you give me write access? It'll be easier if I can push directly.